src/librustc_resolve/lib.rs

```diff
@@ -4268,17 +4269,17 @@ impl<'a, 'crateloader: 'a> Resolver<'a, 'crateloader> {
                 if ident.name == lookup_name && ns == namespace {
                     if filter_fn(name_binding.def()) {
                         // create the path
-                        let mut segms = if self.session.rust_2018() && !in_module_is_extern {
+                        let mut segms = path_segments.clone();
+                        if self.session.rust_2018() {
                             // crate-local absolute paths start with `crate::` in edition 2018
                             // FIXME: may also be stabilized for Rust 2015 (Issues #45477, #44660)
-                            let mut full_segms = vec![
-                                ast::PathSegment::from_ident(keywords::Crate.ident())
-                            ];
-                            full_segms.extend(path_segments.clone());
-                            full_segms
-                        } else {
-                            path_segments.clone()
-                        };
+                            let first_segment_ident = segms[0].ident;
+                            if first_segment_ident.name != "std" {
```

Comment:
So I was looking into this special-casing. I think the reason that I was confused is that I thought that if you had a path like `use foo::bar`, then we would only look for the crate `foo` amongst the extern-prelude crates. But from skimming the code I no longer *quite* think that is the case. I think we'll also search the `-L` directories for that crate -- and this (I believe) is exactly how `use std::foo` works at all. In that case, we can remove the special case here and make instead the change below.

Still, I somehow that wasn't the desired design... so cc @eddyb and @petrochenkov -- how do we get the full list of crates that can be used with an explicit `extern crate`?

Reply:
@eddyb, @petrochenkov -- for context, what we are trying to do is to search the "available" crates for completion, but avoid duplicates from where there is an `extern crate`...